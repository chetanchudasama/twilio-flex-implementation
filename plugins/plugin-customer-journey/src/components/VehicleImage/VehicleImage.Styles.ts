import {
  white,
  borderStyle,
  defaultSnackbarColor,
  darkGrey,
} from "@common/components";
import styled from "react-emotion";

export const VehicleImageStyles = styled("div")({
  width: "100%",
  height: "100%",
  border: borderStyle,
  position: "relative",
  backgroundColor: defaultSnackbarColor,
  borderRadius: "10px 0px 0px 3px",

  ".vehicle-img": {
    width: "100%",
    height: "100%",
    maxWidth: "100%",
    maxHeight: "100%",
    objectFit: "contain",
    cursor: "pointer",
  },
  ".image-count": {
    position: "absolute",
    left: "0",
    top: "0",
    display: "flex",
    justifyContent: "center",
    padding: "10px",
    color: white,
    background: darkGrey,
    cursor: "pointer",
    borderRadius: "10px 0px 5px 0px",
  },
  ".icon": {
    fontSize: "18px",
  },
  ".image-count .icon": {
    paddingRight: "8px",
  },
  ".image-missing": {
    width: "100%",
    height: "100%",
  },
});
