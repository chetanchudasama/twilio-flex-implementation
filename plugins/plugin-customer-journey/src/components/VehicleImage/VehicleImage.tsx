import { ImageCarousel, CarNotFoundIcon } from "@common/components";
import React, { useState, useEffect } from "react";
import CameraAltIcon from "@material-ui/icons/CameraAlt";
import { VehicleImageStyles } from "./VehicleImage.Styles";

export interface VehicleImageProps {
  image: string;
  imageList?: string[];
  imagesCount?: number;
  showImageCount?: boolean;
}

export const VehicleImage: React.FC<VehicleImageProps> = ({
  image,
  imageList,
  imagesCount,
  showImageCount,
}) => {
  const [isImageUrlValid, setIsImageUrlValid] = useState(true);
  const [showCarousel, setShowCarousel] = useState<boolean>(false);

  const hasImages = imageList && imageList.length > 0;

  useEffect(() => {
    const imageExists = (
      url: string,
      callback: (isImageExists: boolean) => void
    ) => {
      const img: HTMLImageElement = new Image();
      img.onload = () => {
        callback(true);
      };
      img.onerror = () => {
        callback(false);
      };
      img.src = url;
    };

    if (image) {
      imageExists(image, (isImageExists) => {
        setIsImageUrlValid(isImageExists);
      });
    } else {
      setIsImageUrlValid(false);
    }
  }, [image]);

  return (
    <>
      <VehicleImageStyles
        onClick={() =>
          isImageUrlValid && hasImages ? setShowCarousel(true) : {}
        }
      >
        {image && isImageUrlValid ? (
          <>
            <img src={image} className="vehicle-img" alt="" />
            {isImageUrlValid && showImageCount && (
              <div className="image-count">
                <CameraAltIcon className="icon" />
                {imagesCount ?? 0}
              </div>
            )}
          </>
        ) : (
          <div className="image-missing">
            <CarNotFoundIcon />
          </div>
        )}
      </VehicleImageStyles>
      {showCarousel && hasImages && (
        <ImageCarousel
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          images={imageList!}
          showCarousel={showCarousel}
          handleClose={() => setShowCarousel(false)}
        />
      )}
    </>
  );
};
