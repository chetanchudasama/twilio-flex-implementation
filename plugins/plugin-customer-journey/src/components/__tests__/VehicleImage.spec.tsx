import React from "react";
import { mount } from "enzyme";
import { create, act } from "react-test-renderer";
import { VehicleImage, VehicleImageProps } from "../VehicleImage/VehicleImage";

const props: VehicleImageProps = {
  image: "ImageUrl",
  imageList: ["ImageUrl"],
  imagesCount: 1,
  showImageCount: true,
};

jest.mock("@common/components", () => {
  const components = jest.requireActual("@common/components");
  return {
    components,
    ImageCarousel: () => {
      return <div className="image-carousel">Image Carousel</div>;
    },
    CarNotFoundIcon: () => {
      return <div className="car-not-found">Car not found</div>;
    },
  };
});

describe("VehicleImage", () => {
  it("renders correctly on load", async () => {
    const wrapper = create(<VehicleImage {...props} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("when value of image prop is null, should render Car not found image icon", async () => {
    props.image = "";
    const wrapper = create(<VehicleImage {...props} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("should renders prop values correctly", async () => {
    props.image = "ImageUrl";
    const wrapper = mount(<VehicleImage {...props} />);

    expect(wrapper.find(".vehicle-img").exists()).toBeTruthy();
    expect(wrapper.find(".image-count").exists()).toBeTruthy();
    expect(wrapper.find(".image-missing").exists()).toBeFalsy();
  });

  it("when value of prop showImageCount is false, UI should render accordingly", async () => {
    props.showImageCount = false;
    const wrapper = mount(<VehicleImage {...props} />);

    expect(wrapper.find(".image-count").exists()).toBeFalsy();
  });

  it("should render ImageCarousel component, when vehicle image is clicked", async () => {
    const wrapper = create(<VehicleImage {...props} />).root;

    act(() => {
      wrapper.findAllByType("div")[0].props.onClick();
    });

    expect(
      wrapper.findAllByProps({ className: "image-carousel" })
    ).toHaveLength(1);
  });

  it("should not render ImageCarousel component, when value of imageList is null", async () => {
    props.imageList = [];
    const wrapper = create(<VehicleImage {...props} />).root;

    act(() => {
      wrapper.findAllByType("div")[0].props.onClick();
    });

    expect(
      wrapper.findAllByProps({ className: "image-carousel" })
    ).toHaveLength(0);
  });
});
