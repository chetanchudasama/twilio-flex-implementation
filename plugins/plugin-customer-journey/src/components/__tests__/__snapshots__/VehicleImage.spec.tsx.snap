// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleImage renders correctly on load 1`] = `
<div
  className="css-jk2enx"
  onClick={[Function]}
>
  <img
    alt=""
    className="vehicle-img"
    src="ImageUrl"
  />
  <div
    className="image-count"
  >
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-1 icon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <circle
        cx="12"
        cy="12"
        r="3.2"
      />
      <path
        d="M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
    1
  </div>
</div>
`;

exports[`VehicleImage when value of image prop is null, should render Car not found image icon 1`] = `
<div
  className="css-jk2enx"
  onClick={[Function]}
>
  <div
    className="image-missing"
  >
    <div
      className="car-not-found"
    >
      Car not found
    </div>
  </div>
</div>
`;
